Fixed #35813 -- Made migrations track all changes to unmanaged models.

hannylicious opened the pull request:
#### Trac ticket number

ticket-35813

#### Branch description
Unmanaged models had no history of alterations being done to them. This PR fixes that issue so that any changes made to unmanaged models will be reflect in the same fashion as managed models (which is indicated as the expected behavior per the docs)

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [x] I have added or updated relevant docs, including release notes if applicable.
- [x] I have attached screenshots in both light and dark modes for any UI changes.

hannylicious committed:
Fixed #35813 -- Made migrations track all changes to unmanaged models.

Previously, only create and delete operations were generated.

charettes reviewed (commented):
Hello @hannylicious!

The patch seems to be on the right way but I suspect that more fundamental changes will be required to have the auto-detector keep track of all model changes.

If we are going to make this change it should not only detect field changes on unmanaged models for all kind of changes (model/field renames, constraint/index additions/alterations).

I suspect the approach we'll want to have here is to stop partitioning model keys by managed or not by getting of `kept_unmanaged_keys` and `old_unmanaged_keys`.

charettes commented:
From some local test removing all the unmanaged model specialization from the autodetector passes the full suite and the test you added so that seems to support the idea that it's an avenue worth exploring.

hannylicious commented:
> Hello @hannylicious!
> 
> The patch seems to be on the right way but I suspect that more fundamental changes will be required to have the auto-detector keep track of all model changes.
> 
> If we are going to make this change it should not only detect field changes on unmanaged models for all kind of changes (model/field renames, constraint/index additions/alterations).
> 
> I suspect the approach we'll want to have here is to stop partitioning model keys by managed or not by getting of `kept_unmanaged_keys` and `old_unmanaged_keys`.

I agree. However there were a couple test that began to break when I applied that methodology, so I decided to rework things and go with this approach. I'll take another crack at it!

hannylicious commented:
@charettes - in reviewing your changes compared to what I had originally, I think the reason I saw test failures was that I had not fully implemented the changes in all appropriate places in the file. Thanks for your local diff - I've went ahead and went forward using that as a guide, updated the PR and added a number of additional tests for various scenarios to ensure things are getting treated appropriately. 

I imagine there are probably a lot more edge cases that could potentially be added to the test suite, but I think the tests added help to cover a lot of the main use cases and further ensure both unmanaged/managed migrations get treated equally in order to comply with the ideas Django has set forth in the docs.

I do wonder if having an option to 'not include unmanaged models in migrations' would be of benefit (I think it would be); but that's a discussion for a different PR all together.

charettes commented:
Thanks for the adjustments and the additional tests ⭐ 

> I do wonder if having an option to 'not include unmanaged models in migrations' would be of benefit (I think it would be); but that's a discussion for a different PR all together.

I also think this should be brought up as a distinct issue on the form, I do fear complications around what happens if you stop passing the option mid-project. I think that concurrent work such as #18445 (ticket-35656) could address your issue though as you would _only_ have to define your own `autodetector_class` that purge unmanaged model entries from `new_model_keys`.

kevingill1966 commented:
I reviewed this ticket.

I followed the checklist as per: https://docs.djangoproject.com/en/dev/internals/contributing/new-contributors/#guidelines

1. The patch fixes the problem.
2. The tests fail without the patch, and succeed with the patch.
3. The patch passes the checkers black, sort, flake8, and blacken-doc

### Packaging Issues

1. The PR does not provide a single commit. There are two commits. The commit message does not follow the guidelines as per:  https://docs.djangoproject.com/en/dev/internals/contributing/committing-code/#committing-guidelines

2. The PR is not documented in the release documentation.

### Conclusion

This PR works well. The patch should be merged. It just needs to be prepared for the merge process.

hannylicious commented:
@kevingill1966 I believe I have updated everything with the changes you requested. Let me know if you see any other things that might need changing!

nessita commented:
> @nessita thanks for the heads up! I didn't realize I needed to update the ticket. I have done so, as well as updated the pr to reflect the most recent django code.

Thank you for doing that! A minor clarification: a PR author should never mark their own ticket as "Ready for checkin", only reviewers would do that once the work is actually ready to be merged.

blighj reviewed (changes requested):
Thanks again for doing this, shame it fell by the wayside. I've retested it against the latest master and the issue is still there and this patch still fixes it.

To move it forward the PR needs to be rebased and as I suggest the release docs now need to be updated.. Once that is done, we can update the trac ticket so "Needs Documentation" is unset and it will be back in the review queue and maybe it will have a chance of a review for 6.1 or 6.2.

blighj commented:
I've updated that earlier, so this ticket is in the queue again, but it's a very busy at the minute, there are almost a 100 tickets in the queue, normally it is a third of that. It's shot up in the last week or two.

charettes commented on tests/migrations/test_autodetector.py:
```diff
@@ -4019,8 +4107,45 @@ def test_proxy_to_mti_with_fk_to_proxy_proxy(self):
         self.assertEqual(fk_field.remote_field.model, "testapp.AAuthorProxyProxy")
 
+    def test_unmanaged_add_constraints(self):
+        """Test change detection of new constraints."""
```
We usually don't include docstrings that are redundant with the test name as don't provide much value. I see that this test module doesn't necessarily follows this convention though.

timgraham replied:
I guess all docstrings should have "on an unmanaged model." (only one does)

timgraham commented:
I'd suggest to update the PR description to "Fixed #35813 -- Made migrations track all changes to unmanaged models." The second line of the commit message could elaborate: "Previously, only create and delete operations were generated." There might be time to get this merged as a bug fix for Django 6.1 before it reaches "beta" status next month. 

(The fix is of interest to me as I'm experimenting with making Django MongoDB Backend's `EmbeddedModel` unmanaged as a way of preventing `SchemaEditor` from operating on it, as well as a way of detecting embedded models in the autodetector. I believe I need `AddField`, `AlterField`, etc. operations for embedded models so that their model state is updated with added/removed/changed fields.)

hannylicious commented:
Not sure if it's too late for review into 6.1 - but I've updated this PR in order to better reflect what it does as a feature in 6.1.

I am grateful for the help/guidance offered in this thread! This has been a great way to learn more on helping to make Django better.

timgraham commented on django/db/migrations/autodetector.py:
"unmanaged" could also be removed from this docstring.

timgraham commented on docs/releases/6.1.txt:
```diff
@@ -274,6 +274,11 @@ Management Commands
   loading fixtures.
 
+* The :djadmin:`makemigrations` command now tracks all changes to unmanaged
+  models, including field additions, removals, alterations, constraints and
```
constraints, (add comma)

timgraham commented on docs/releases/6.1.txt:
```diff
@@ -274,6 +274,11 @@ Management Commands
+* The :djadmin:`makemigrations` command now tracks all changes to unmanaged
+  models, including field additions, removals, alterations, constraints and
+  model renames. After upgrading, you will see new migrations detected for
+  unmanaged models that have changed.
```
"have changed since their creation."

timgraham commented on tests/migrations/test_autodetector.py:
```diff
@@ -4019,8 +4107,45 @@ def test_proxy_to_mti_with_fk_to_proxy_proxy(self):
+        self.assertOperationAttributes(changes, "testapp", 0, 0, name="name")
+
+    def test_unmanaged_alter_field(self):
+        """Tests autodetection of new fields on an unmanaged model."""
```
new -> altered

timgraham reviewed (commented):
The PR title still uses the old commit message.

hannylicious commented:
I've made the requested changes (which were great, thanks Tim!)

sarahboyce reviewed (commented):
Hello hello 😁 thank you all for the work on this PR, it looks really good

I am concerned that, due to the changes to unmanaged models not being detected previously, some folks may have been able to change their models in ways that wouldn't normally be allowed. Then when trying to upgrade now, they would struggle

An example upgrade paper cut is adding fields without defaults. Someone upgrading now may be bombarded with a lot of fields they need to provide one off defaults for. Though this isn't an example that I am too worried about

I was specifically concerned about adding and changing `CompositePrimaryKey`s as this is something the docs say we do not support, and is a real example within djangoproject.com, but when testing this wasn't an issue in practice

Are there any examples folks can think of where that could prove to be an issue? Perhaps we would need a section within the release notes that would have space for upgrade guidelines if required. We may need to clearly explain the value as I can see folks being like "this has been work for me and I don't see the value".

sarahboyce reviewed (approved):
I don't think my concern was well founded 😁 
Thank you all

jacobtylerwalls commented:
Thanks everyone for seeing this through!

> I don't think my concern was well founded 😁

Not giving yourself enough credit there! I do wonder if we should consider a solution for:

> An example upgrade paper cut is adding fields without defaults. Someone upgrading now may be bombarded with a lot of fields they need to provide one off defaults for. 

I just experienced that in a test project. Do we want the autodetector to complain in this situation? On first impression, if a model is _unmanaged_, then it isn't really "impossible" to add a non-nullable default for a field. Someone else is managing that table.

What do folks think?

jacobtylerwalls commented:
Opened ticket-37224 to discuss this ☝️

hannylicious commented:
Great ticket to open @jacobtylerwalls and thank you @sarahboyce for initially pointing out those 'sharp edges' that come with this change. I think your initial thoughts have merit and are important considerations! 

I think the autodetector probably shouldn't complain, because if it's an unmanaged model, while we want to see changes, I don't think we want it asking us for default information. It's unmanaged for a reason, so I think it's safe to assume we want it to not complain, just accept what we tell it as true.
